draft-ietf-httpbis-priority.md

```diff
@@ -320,41 +343,37 @@ browser would send a reprioritization frame with the priority field value
-using the Priority header field.  Modifying this behavior would require a
-semantic change to the protocol, but this is avoided by restricting the
-stream on which a PRIORITY_UPDATE frame can be sent. In HTTP/2 the frame
-is on stream zero and in HTTP/3 it is sent on the control stream
-({{!I-D.ietf-quic-http}}, Section 6.2.1).
-
-This document specifies a new PRIORITY_UPDATE frame type for HTTP/2
-({{!RFC7540}}) and HTTP/3 ({{!I-D.ietf-quic-http}}) which enables
-reprioritization. It carries updated priority parameters and references the
-target of the reprioritization based on a version-specific identifier; in
-HTTP/2 this is the Stream ID, in HTTP/3 this is either the Stream ID or Push ID.
-
-Unlike the header field, the reprioritization frame is a hop-by-hop signal.
+transitions to a state that prevents the client from sending additional frames
+on the stream. Therefore, a client cannot reprioritize a response by sending
+additional metadata (e.g., trailer fields) or a PRIORITY_UPDATE frame on the
+request stream. We avoid interoperability failure by restricting PRIORITY_UPDATE
+frames to stream zero or the control stream, with each such frame explicitly
+identifying the response to which the new priority applies.
 
 ## HTTP/2 PRIORITY_UPDATE Frame {#h2-update-frame}
 
-The HTTP/2 PRIORITY_UPDATE frame (type=0xF) carries the stream ID of the
-response that is being reprioritized, and the updated priority in ASCII text,
-using the same representation as that of the Priority header field value.
+The HTTP/2 PRIORITY_UPDATE frame (type=0xF) is used by clients to signal the
+initial priority of a response, or to reprioritize a response or push stream. It
+carries the stream ID of the response and the priority in ASCII text, using the
+same representation as that of the Priority header field value.
 
 The Stream Identifier field ({{!RFC7540}}, Section 4.1) in the PRIORITY_UPDATE
-frame header MUST be zero (0x0).
+frame header MUST be zero (0x0). Receiving a PRIORITY_UPDATE frame with a field
+of any other value MUST be treated as a connection error of type PROTOCOL_ERROR.
 
 ~~~ drawing
-  0                   1                   2                   3
-  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
- +---------------------------------------------------------------+
- |R|                        Stream ID (31)                       |
- +---------------------------------------------------------------+
- |                   Priority Field Value (*)                  ...
- +---------------------------------------------------------------+
+HTTP/2 PRIORITY_UPDATE Frame {
+  Length (24),
+  Type (8) = 0x10,
+  Flags (8),
+  Reserved (1),
+  Stream Identifier (31),
+  R (1),
+  Prioritized Stream ID (31),
+  Priority Field Value (..),
+}
 ~~~
-{: #fig-h2-reprioritization-frame title="HTTP/2 PRIORITY_UPDATE Frame Payload"}
+{: #fig-h2-reprioritization-frame title="HTTP/2 PRIORITY_UPDATE Frame"}
```

Comment:
Seeing the definition of the frame _headers_ for H2 was surprising to me, though I understand that it reflects the change to the HTTP/3 draft. Maybe I should ask if the change in HTTP/3 is a good thing.

Reply:
As a personal opinion I prefer to see the frame in whole because it avoids having to cross reference to the frame headers, which are defined in some other document. 

But I can drop them if people want. I see that as an editorial change so perhaps we can avoid blocking this and fix up later if needs be.

Reply:
I agree that this is an editorial issue, and that we need to ask the WG to resolve (because it is about the convention covering all H2 extensions, not just this one).

Reply:
FWIW, ALTSVC is the only extension to have reached RFC status previously.  It [does not include](https://tools.ietf.org/html/rfc7838#section-4) the frame header; only the payload.

Reply:
You forgot ORIGIN frame https://tools.ietf.org/html/rfc8336#section-2.1 which also specifies the frame payload.

Now that H2 and H3 have different frame format definitions, I think extensions applicable to both will hit this problem. I'd like to figure what people think is the most acceptable way to present things.

Reply:
One option would be to write documents that update RFC 7838 and RFC 8336 to use the new format and define the H3 frame at the same time...

Reply:
True, I try to block that one out.

They have different frame format definitions, sure; the question is whether the payload needs to vary between H2/H3.  I think the real things you need to specify are:

- What stream it's on (part of the header in H2, intrinsic property in H3)
  - Control stream vs. stream zero language plays into this
- Flags if used in H2; since H3 doesn't have flags, this might lead to a payload difference
- Payload

The simplest path for dual-version extensions is not to use flags and only define the payload.

Reply:
Taking the feedback received here and elsewhere, I'm reverted the diagram back to our traditional format.